### PR TITLE
add RemoveRange for in-memory connector

### PR DESCRIPTION
**Pull request by klnusbaum**
We'd like to add the ability to remove a range of values from DOSA. This adds that ability to the in-memory connector.

### CODE REVIEW

File: connectors/memory/memory.go [hunk 1]
```diff
@@ -326,13 +334,51 @@ func (c *Connector) Remove(_ context.Context, ei *dosa.EntityInfo, values map[st
...
+func (c *Connector) RemoveRange(_ context.Context, ei *dosa.EntityInfo, columnConditions map[string][]*dosa.Condition) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	partitionRange := c.findRange(ei, columnConditions)
+	if partitionRange == nil {
+		return nil
+	}
+
+	partitionRef := partitionRange.entityRef[partitionRange.partitionKey]
```
Comment (klnusbaum) [inline connectors/memory/memory.go]: This all feels really bad, would love a better way to do this with out causing
```
entityRef[partitionKey] = append(partitionRef[:start], partitionRef[end+1:]...)
```
to be really long

Comment (rkuris) [inline connectors/memory/memory.go]: Did you look at making methods on partitionRange that do the heavy lifting? Something like:
```
func (p *partitionRange) deleteRange(er *[]map[string]dosa.FieldValue) {
    er = append(er[:p.start], er[p.end:]...)
```
The caller would then do:
```
partitionRange.deleteRange(&entityRef[partitionKey])
```

File: connectors/memory/memory_test.go [hunk 2]
```diff
@@ -239,6 +239,53 @@ func TestConnector_Remove(t *testing.T) {
...
+	const idcount = 10
+	sut := NewConnector()
+
+	// remove with no data
+	err := sut.RemoveRange(context.TODO(), clusteredEi, map[string][]*dosa.Condition{
+		"f1": {{Op: dosa.Eq, Value: dosa.FieldValue("data")}},
+	})
+	assert.NoError(t, err)
+
+	// insert some data into data/i/uuid with a random set of uuids
```
Comment (rkuris) [inline connectors/memory/memory_test.go]: nit: copy/paste from another test ;)

File: connectors/memory/memory.go (hunk 1 above)
Comment (rkuris) [inline connectors/memory/memory.go]: If you don't like the pointer approach, you can return the updated value with the range deleted (I think the pointer approach is more elegant though)

File: connectors/memory/memory_test.go [hunk 3]
```diff
@@ -239,6 +239,53 @@ func TestConnector_Remove(t *testing.T) {
...
+	for i, x := range data {
+		assert.Equal(t, x["c1"], int64(i))
+	}
+
+	// ensure all the uuids greater than the uuid at the midpoint are removed.
+	data, _, err = sut.Range(context.TODO(), clusteredEi, map[string][]*dosa.Condition{
+		"f1": {{Op: dosa.Eq, Value: dosa.FieldValue("data")}},
+		"c1": {{Op: dosa.Gt, Value: dosa.FieldValue(int64(4))}},
+	}, dosa.All(), "", 200)
+	t.Logf("Length of data is: %d\n", len(data))
```
Comment (rkuris) [inline connectors/memory/memory_test.go]: remove this debug code

File: connectors/memory/memory_test.go [hunk 4]
```diff
@@ -239,6 +239,53 @@ func TestConnector_Remove(t *testing.T) {
...
+	assert.Error(t, err)
+	assert.True(t, dosa.ErrorIsNotFound(err))
+}
```
Comment (rkuris) [inline connectors/memory/memory_test.go]: missing test: remove everything from a partitionKey

File: connectors/memory/memory.go [hunk 5]
```diff
@@ -326,13 +350,43 @@ func (c *Connector) Remove(_ context.Context, ei *dosa.EntityInfo, values map[st
...
 }
 
+// RemoveRange removes all of the elements in the range specified by the entity info and the column conditions.
+func (c *Connector) RemoveRange(_ context.Context, ei *dosa.EntityInfo, columnConditions map[string][]*dosa.Condition) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	partitionRange := c.findRange(ei, columnConditions)
+	if partitionRange == nil {
+		return nil
```
Comment (rkuris) [inline connectors/memory/memory.go]: Nit: complexity: if partitionRange!=nil { pr.delete()}

**Pull request merged**